### PR TITLE
Client Encryption: Removes retry behavior for point operations during client encryption policy change.

### PR DESCRIPTION
**Pull request by kr-santosh**
## Description

This PR brings in the following changes and fixes 

- changes the retry logic where few operations (create/read/upsert etc) were retried upon failure due to policy changes. With this PR this retry has been removed and we just refresh the policy. This is to align with and have a uniform behavior across, since few operations like batch, feed cannot be retried, and the stream dispose behavior (when operation fails) seems to be inconsistent with bulk.

- adds fix for patch operation to pick up (and refresh the cache with) the correct encryption policy when the client has cached incorrect/stale policy which can happen when container is deleted and recreated with same id but with a different encryption policy.

## Type of change

- [] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs [hunk 1]
```diff
@@ -1235,6 +1213,151 @@ private static ItemRequestOptions GetClonedItemRequestOptions(ItemRequestOptions
...
+            responseMessage.Content = await EncryptionProcessor.DecryptAsync(
+                responseMessage.Content,
+                encryptionSettings,
+                diagnosticsContext,
+                cancellationToken);
+
+            return responseMessage;
+        }
+
+        private async Task<List<PatchOperation>> EncryptPatchItemsAsync(
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs:1283]: This seems to be a duplicate of EncryptPatchOperationsAsync. Please refactor.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 2]
```diff
@@ -1292,6 +1292,123 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteQuery()
...
+
+            clientEncryptionPolicy = new ClientEncryptionPolicy(paths);
+
+            containerProperties = new ContainerProperties(encryptionContainerToDelete.Id, "/PK") { ClientEncryptionPolicy = clientEncryptionPolicy };
+
+            ContainerResponse containerResponse = await database.CreateContainerAsync(containerProperties, 400);
+
+            TestDoc docPostPatching = await MdeEncryptionTests.MdeCreateItemAsync(encryptionContainerToDelete);
+
+
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1381]: nit: extra line

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 3]
```diff
@@ -1292,6 +1292,123 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteQuery()
...
+            CosmosClient otherClient = TestCommon.CreateCosmosClient(builder => builder
+                .Build());
+
+            CosmosClient otherEncryptionClient = otherClient.WithEncryption(new TestEncryptionKeyStoreProvider());
+            Database otherDatabase = otherEncryptionClient.GetDatabase(MdeEncryptionTests.database.Id);
+
+            Container otherEncryptionContainer = otherDatabase.GetContainer(encryptionContainerToDelete.Id);
+
+            await MdeEncryptionTests.MdeCreateItemAsync(otherEncryptionContainer);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1332]: why do we need this?

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 4]
```diff
@@ -1292,6 +1292,123 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteQuery()
...
+                PatchOperation.Replace("/Sensitive_DateFormat", docPostPatching.Sensitive_DateFormat),
+                PatchOperation.Replace("/Sensitive_DecimalFormat", docPostPatching.Sensitive_DecimalFormat),
+                PatchOperation.Set("/Sensitive_IntArray/1", docPostPatching.Sensitive_IntArray[1]),
+                PatchOperation.Set("/Sensitive_IntMultiDimArray/1/0", docPostPatching.Sensitive_IntMultiDimArray[1,0]),
+                PatchOperation.Replace("/Sensitive_IntFormat", docPostPatching.Sensitive_IntFormat),
+                PatchOperation.Replace("/Sensitive_BoolFormat", docPostPatching.Sensitive_BoolFormat),
+                PatchOperation.Replace("/Sensitive_FloatFormat", docPostPatching.Sensitive_FloatFormat),
+            };
+
+            await MdeEncryptionTests.MdePatchItemAsync(otherEncryptionContainer, patchOperations, docPostPatching, HttpStatusCode.OK);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1409]: Please validate that the expected paths (as per new policy) are encrypted.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 5]
```diff
@@ -1292,6 +1292,123 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteQuery()
...
+            // Maximum 10 operations at a time (current limit)
+            List<PatchOperation> patchOperations = new List<PatchOperation>
+            {
+                PatchOperation.Increment("/NonSensitiveInt", 1),
+                PatchOperation.Replace("/NonSensitive", docPostPatching.NonSensitive),
+                PatchOperation.Replace("/Sensitive_StringFormat", docPostPatching.Sensitive_StringFormat),
+                PatchOperation.Replace("/Sensitive_DateFormat", docPostPatching.Sensitive_DateFormat),
+                PatchOperation.Replace("/Sensitive_DecimalFormat", docPostPatching.Sensitive_DecimalFormat),
+                PatchOperation.Set("/Sensitive_IntArray/1", docPostPatching.Sensitive_IntArray[1]),
+                PatchOperation.Set("/Sensitive_IntMultiDimArray/1/0", docPostPatching.Sensitive_IntMultiDimArray[1,0]),
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1403]: test Sensitive_NestedObjectFormatL1 path instead

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs (hunk 1 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs]: ah...this was caused due to a merge conflict ..fixed it..

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 3 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1332]: just an operations before the container gets deleted, results in caching of policy. Since am not calling init.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 4 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1409]: Yes currently this auto validated since we call decrypt to read out the response and the response would fail if the policy is incorrect. since the item was created from the client.
Also I have work item, to add to the validation which is currently missing in the test suite for all test to verify if the paths were actually encrypted. Will add that in soon.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 5 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1403]: the idea was to not use the same policy paths. This should be good enough.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 3 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1332]: Makes sense. Would be good if you can add this as a comment.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 5 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1403]: The reason I asked was because Sensitive_NestedObjectFormatL1 path is part of policy which gets replaced. So we can verify that it doesnt get encrypted as per new policy.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 4 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1409]: The point is to verify that the correct policy was picked to encrypt / decrypt. Not sure we validate that at the moment.

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1409]: yes...like I said.. I will add a helper function to check the same for all test cases soon.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 5 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1403]: sure.

**Review by anujtoshniwal — APPROVED** [review]

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 6]
```diff
@@ -1272,13 +1270,13 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteQuery()
                 Assert.Fail("ValidateQueryResultAsync should have failed. ");
             }
             catch(CosmosException ex)
-            {
+            {
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1273]: nit: remove extra spaces

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 7]
```diff
@@ -1169,9 +1169,7 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteTransactionB
             }
             catch(CosmosException ex)
             {
-                Assert.IsNotNull(ex);
-                // Github issue tracking fix: https://github.com/Azure/azure-cosmos-dotnet-v3/issues/2714
-                // Assert.AreEqual("Operation has failed due to a possible mismatch in Client Encryption Policy configured on the container. Please refer to https://aka.ms/CosmosClientEncryption for more details. ", ex.Message);
+                Assert.IsNotNull(ex.Message);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1172]: Can we validate the expected message?

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 8]
```diff
@@ -1272,13 +1270,13 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteQuery()
...
             catch(CosmosException ex)
-            {
+            {                
                 if (ex.SubStatusCode != 1024)
                 {
                     Assert.Fail("Query should have failed. ");
                 }
 
-                Assert.IsTrue(ex.Message.Contains("Operation has failed due to a possible mismatch in Client Encryption Policy configured on the container. Please refer to https://aka.ms/CosmosClientEncryption for more details. "));
+                Assert.IsNotNull(ex.Message);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1279]: Why this change?

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1342]: For some reason the exception message was getting truncated in gates causing this failure. Hence removed this. Need to investigate.

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs [hunk 9]
```diff
@@ -85,22 +89,35 @@ public async Task<AeadAes256CbcHmac256EncryptionAlgorithm> BuildEncryptionAlgori
...
+                    return protectedDataEncryptionKey;
+                }
+                finally
+                {
+                    EncryptionKeyCacheSema.Release(1);
+                }
+            }
 
-            return protectedDataEncryptionKey;
+            return null;
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs:120]: Throw exception indicating timeout? Returning null here would lead to NullRefException at a later point in the code flow?

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 10]
```diff
@@ -1045,6 +1045,21 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteWithBulk()
...
+                Assert.Fail("Create operation should have failed.");
+            }
+            catch (CosmosException ex)
+            {
+                if (ex.SubStatusCode != 1024)
+                {
+                    Assert.Fail("Create operation should have failed with 1024 SubStatusCode. ");
+                }
+
+                Assert.IsNotNull(ex.Message);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1060]: nit:
Assert.AreEqual(ex.SubStatusCode, 1024, "Create operation should have failed with 1024 SubStatusCode.");

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs (hunk 9 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs:120]: we don't timeout here (setting to -1).Failure to acquire a lock should return a null, unless  cancelled.  I think you are right since MDE lib will throw null exception when we pass the null PDEK.

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs [hunk 11]
```diff
@@ -13,9 +13,7 @@ namespace Microsoft.Azure.Cosmos.Encryption
 
     internal sealed class EncryptionSettingForProperty
     {
-        public string ClientEncryptionKeyId { get; }
-
-        public EncryptionType EncryptionType { get; }
+        private static readonly SemaphoreSlim EncryptionKeyCacheSema = new SemaphoreSlim(1, 1);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs:16]: Private variable name should start with lowercase letter.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 8 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1346]: If thats the case, please create a tracking item for the same.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 7 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1224]: This is the exception which the user will get as well. Will be good to make sure there are no surprises and also the message is clear.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 12]
```diff
@@ -1272,13 +1337,13 @@ public async Task EncryptionValidatePolicyRefreshPostContainerDeleteQuery()
                 Assert.Fail("ValidateQueryResultAsync should have failed. ");
             }
             catch(CosmosException ex)
-            {
+            {
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1340]: nit: extra spaces

**Comment by abhijitpai** [comment]
>                 streamPayload.Position = 0;

Still needed?

---
Refers to: Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs:872 in a1cfd32. [](commit_id = a1cfd327a10bbde0522b2f27539a7d69f8cc3d64, deletion_comment = False)

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs [hunk 13]
```diff
@@ -889,18 +874,16 @@ private static ItemRequestOptions GetClonedItemRequestOptions(ItemRequestOptions
...
 
-                // we try to recreate the item with the StreamPayload(to be encrypted) now that the encryptionSettings would have been updated with latest values if any.
-                return await this.CreateItemHelperAsync(
-                       streamPayload,
-                       partitionKey,
-                       clonedRequestOptions,
-                       cancellationToken,
-                       isRetry: true);
+                throw new CosmosException(
+                   "Operation has failed due to a possible mismatch in Client Encryption Policy configured on the container. Retrying can possibly fix the issue. Please refer to https://aka.ms/CosmosClientEncryption for more details. " + responseMessage.ErrorMessage,
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs:882]: >can possibl [](http://example.com/codeflow?start=134&length=11)

"may" instead of "can possibly"

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs [hunk 14]
```diff
@@ -748,6 +730,17 @@ internal sealed class EncryptionContainer : Container
                 cancellationToken: cancellationToken);
         }
 
+        internal static bool CheckIfRequestNeedsARetryPostPolicyRefresh(ResponseMessage responseMessage)
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs:733]: >CheckIfRequestNeedsARetryPostPolicyRefresh [](http://example.com/codeflow?start=29&length=42)

Should this be ThrowIfRequestNeedsARetryPostPolicyRefresh and do the repeated code too?

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs (hunk 11 above)
Comment (abhijitpai) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs:16]: Also please use Semaphore and not Sema

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettings.cs [hunk 15]
```diff
@@ -33,6 +33,11 @@ public static Task<EncryptionSettings> CreateAsync(EncryptionContainer encryptio
 
         public EncryptionSettingForProperty GetEncryptionSettingForProperty(string propertyName)
         {
+            if (string.IsNullOrEmpty(propertyName))
+            {
+                return null;
+            }
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettings.cs:39]: Revert? Let's avoid these kind of implicit contracts.

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs (hunk 14 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionContainer.cs:733]: Sure!

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs (hunk 11 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs:16]: ah. So static readonly variables should  begin with UpperCase :) . Yup will rename it

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs [hunk 16]
```diff
@@ -85,22 +89,35 @@ public async Task<AeadAes256CbcHmac256EncryptionAlgorithm> BuildEncryptionAlgori
...
             EncryptionKeyStoreProvider encryptionKeyStoreProvider,
-            string keyId)
+            string keyId,
+            CancellationToken cancellationToken)
         {
-            KeyEncryptionKey keyEncryptionKey = KeyEncryptionKey.GetOrCreate(
-               clientEncryptionKeyProperties.EncryptionKeyWrapMetadata.Name,
-               clientEncryptionKeyProperties.EncryptionKeyWrapMetadata.Value,
-               encryptionKeyStoreProvider);
+            if (await EncryptionKeyCacheSemaphore.WaitAsync(-1, cancellationToken))
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs:98]: You can just pass the cancellationtoken, no need for the timeout? https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.waitasync?view=net-5.0#System_Threading_SemaphoreSlim_WaitAsync_System_Threading_CancellationToken_

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs [hunk 17]
```diff
@@ -85,22 +89,35 @@ public async Task<AeadAes256CbcHmac256EncryptionAlgorithm> BuildEncryptionAlgori
...
+                    ProtectedDataEncryptionKey protectedDataEncryptionKey = ProtectedDataEncryptionKey.GetOrCreate(
+                        keyId,
+                        keyEncryptionKey,
+                        clientEncryptionKeyProperties.WrappedDataEncryptionKey);
+
+                    return protectedDataEncryptionKey;
+                }
+                finally
+                {
+                    EncryptionKeyCacheSemaphore.Release(1);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs:116]: Releasing 1 is the default behavior, isn't it?

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs (hunk 16 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettingForProperty.cs:98]: yes, again. They exhibit same behavior, just easy to understand...you wait out unless you get a cancel..

File: Directory.Build.props [hunk 18]
```diff
@@ -6,7 +6,7 @@
 		<ClientPreviewSuffixVersion>preview</ClientPreviewSuffixVersion>
 		<DirectVersion>3.23.1</DirectVersion>
 		<EncryptionVersion>1.0.0-previewV18</EncryptionVersion>
-		<CustomEncryptionVersion>1.0.0-previewV2</CustomEncryptionVersion>
```
Comment (ealsur) [inline Directory.Build.props]: We should be a bit more consistent. Encryption uses the previewVXX suffix, SDK uses previewXX (not including 0). Either this is previewV2 (current) or preview2 ? Since this is aligned with encryption, should it follow Encryption pattern?

Comment (kr-santosh) [reply Directory.Build.props]: as discussed offline

**Review by kirankumarkolli — APPROVED** [review]

**Review by abhijitpai — APPROVED** [review]
:shipit:
